### PR TITLE
Static instrumenter logic - a new agent distro

### PR DESCRIPTION
**Pull request by Enkelian**
**Description:**

An alternative to #278.
As discussed, instead of modifying the agent with ByteBuddy, we create a new distro of the agent. The output of `./gradlew :static-instrumenter:agent-instrumenter:assemble` consists of two jars:
* `opentelemetry-static-agent.jar` - used for static instrumentation
* `opentelemetry-no-inst-agent.jar` - needed on application classpath after instrumentation, contains agent's classes with `/inst' and '.classdata` renamed back to plain `.class` files

The new approach also uses the new listener in `HelperInjector`, added here https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/5910
Until it is merged and released, we need to have local dependencies on `opentelemetry-javaagent` and `opentelemetry-muzzle`.

Modules used in the distro:
* `agent-instrumenter` - contains main static agent classes as well  `OpenTelemetryStaticAgent` - a wrapper for `OpenTelemetryAgent` that surrounds `premain`/`agentmain` calls with installing transformers for static instrumentation (`PreTransformer` and `PostTransformer`)
* `agent-extension` - an extension to the agent *that installs a listener in `HelperInjector`) and some additional configuration (that registers a boostrap package)
* `common` - utility classes that need to be accessible from both the agent classes and the static instrumenter main classes

**Existing Issue(s):**

Part of #156

**Testing:**

* unit test of logger
* integration test of the static agent on a sample application

**Documentation:**

* updated readme

**Outstanding items:**

* using test app from  #313 in the integration test
* better way to create `opentelemetry-no-inst-agent.jar` or implementing a solution that makes it obsolete
* removing local dependencies once new agent is released
* configurable logger levels

### CODE REVIEW

File: static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java:30]: Is [this](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/static-instrumenter/test-app) where the app code comes from? Can you add a comment explaining that?

File: static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java:48]: Can you assert on the exit code? To make sure that the instrumentation process has finished successfully

File: static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java:35]: Nit: you can just concatenate strings

File: static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java:63]: Same here, can you assert on the exit code?

File: static-instrumenter/agent-instrumenter/src/main/java/io/opentelemetry/contrib/staticinstrumenter/agent/main/ClassArchive.java
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/src/main/java/io/opentelemetry/contrib/staticinstrumenter/agent/main/ClassArchive.java:77]: 👍

File: static-instrumenter/agent-instrumenter/src/main/java/io/opentelemetry/contrib/staticinstrumenter/agent/main/Main.java
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/src/main/java/io/opentelemetry/contrib/staticinstrumenter/agent/main/Main.java:51]: That can probably stay here

File: static-instrumenter/common/build.gradle.kts
Comment (mateuszrzeszutek) [inline static-instrumenter/common/build.gradle.kts:5]: It's not the extension module, it's the common bootstrap classes module

**Review by mateuszrzeszutek — COMMENTED** [review]
Looks really promising, thanks! 🚀

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (Enkelian) [reply static-instrumenter/agent-instrumenter/build.gradle.kts:136]: It's because `io.opentelemetry.instrumentation.api.config.Config` is used in `agent-extension`, eg [here](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/319/files#diff-ca9deb6e10bef4c5710b60496a0519716cf8b45356a7b8ff84aeab89ce05e1c7R23-R24)

File: static-instrumenter/README.md
Comment (Enkelian) [reply static-instrumenter/README.md:18]: I'm not yet sure how to approach that, but I'll be working on it

File: static-instrumenter/agent-extension/build.gradle.kts
Comment (Enkelian) [reply static-instrumenter/agent-extension/build.gradle.kts:21]: True, thanks!

Comment (Enkelian) [reply static-instrumenter/agent-extension/build.gradle.kts:21]: Sure!

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (mateuszrzeszutek) [reply static-instrumenter/agent-instrumenter/build.gradle.kts:136]: Can you add a comment explaining why we're relocating? (i.e. because of instrumentation-api)

File: static-instrumenter/README.md
Comment (mateuszrzeszutek) [reply static-instrumenter/README.md:18]: Sure, that's just a thing that could be done later

File: static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java
Comment (Enkelian) [reply static-instrumenter/agent-instrumenter/src/integrationTest/java/JarTest.java:65]: done

File: static-instrumenter/agent-extension/build.gradle.kts
Comment (Enkelian) [reply static-instrumenter/agent-extension/build.gradle.kts:19]: Tried using SNAPSHOT, seems like the changes are not available yet, but I'll change it as soon as it's possible

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (Enkelian) [reply static-instrumenter/agent-instrumenter/build.gradle.kts:40]: added TODO

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (Enkelian) [reply static-instrumenter/agent-instrumenter/build.gradle.kts:70]: Done. On a similar note, I think we should change the name of `agent-instrumenter` to `static-agent`, since it no longer instruments the agent but provides a distro of the agent. WDYT?

File: static-instrumenter/agent-extension/build.gradle.kts
Comment (mateuszrzeszutek) [reply static-instrumenter/agent-extension/build.gradle.kts:19]: Hmm, you may need to add the snapshot repo to the gradle config:
```kotlin
repositories {
  mavenCentral()
  maven {
    url = uri("https://oss.sonatype.org/content/repositories/snapshots")
  }
}
```

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (mateuszrzeszutek) [reply static-instrumenter/agent-instrumenter/build.gradle.kts:70]: Yeah, it's a good call 👍

File: static-instrumenter/agent-extension/src/main/java/io/opentelemetry/contrib/staticinstrumenter/extension/AdditionalClassesInjectorListenerInstaller.java
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-extension/src/main/java/io/opentelemetry/contrib/staticinstrumenter/extension/AdditionalClassesInjectorListenerInstaller.java:16]: Hmm, remove the `Installer` word?

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/build.gradle.kts:116]:
```suggestion
  // we depend on opentelemetry-instrumentation-api in agent-extension so we need to relocate its usage
```

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/build.gradle.kts:117]:
```suggestion
  relocate("io.opentelemetry.instrumentation.api", "io.opentelemetry.javaagent.shaded.instrumentation.api")
```

File: static-instrumenter/agent-extension/build.gradle.kts
Comment (Enkelian) [reply static-instrumenter/agent-extension/build.gradle.kts:19]: Yeah, I tried it exactly like this and it seemed like the `HelperInjectorListener` wasn't available yet. I see that there is a new version since I've tried so it should be working now

Comment (Enkelian) [reply static-instrumenter/agent-extension/build.gradle.kts:19]: Still no luck, gradle seems to always download the 1.13.0-alpha version, so I'll have to look into it a bit more

Comment (mateuszrzeszutek) [reply static-instrumenter/agent-extension/build.gradle.kts:19]: Hmm, did you change the bom version in `dependencyManagement`? Does it still not work?

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/build.gradle.kts:69]: Is this still a TODO? Doesn't the task below do just that?

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/build.gradle.kts:115]: Nit: this and the block from line 120 can be put inside the huge `tasks {}` block above

**Review by mateuszrzeszutek — APPROVED** [review]
LGTM 👍 

Before merging, let's try to fix these SNAPSHOTs one more time, it seems to be a better choice than adding jars.

I really like that `static-agent` idea, but that can be done in a separate PR.

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (Enkelian) [reply static-instrumenter/agent-instrumenter/build.gradle.kts:69]: the resulting jar still contains inst (I couldn't configure CopySpec to remove the files), so I'll be working on removing them in a future PR

File: static-instrumenter/agent-extension/build.gradle.kts
Comment (Enkelian) [reply static-instrumenter/agent-extension/build.gradle.kts]: I tried to force using SNAPSHOT only in the necessary dependencies. Setting it directly in `dependencyManagement` finally helped, thanks!

Comment (Enkelian) [reply static-instrumenter/agent-extension/build.gradle.kts]: It builds locally, but CI seems to not search for it in the right repo

File: static-instrumenter/agent-extension/build.gradle.kts
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-extension/build.gradle.kts:3]: Nit: this seems to be unused in this module

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (mateuszrzeszutek) [inline static-instrumenter/agent-instrumenter/build.gradle.kts:30]: The versions should already be set by the `dependencyManagement` -- can you try removing it and checking if it works?

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (mateuszrzeszutek) [reply static-instrumenter/agent-instrumenter/build.gradle.kts:68]: Cool, I'm fine with that 👍

**Review by mateuszrzeszutek — APPROVED** [review]

File: static-instrumenter/agent-extension/build.gradle.kts
Comment (mateuszrzeszutek) [reply static-instrumenter/agent-extension/build.gradle.kts]: Hmm, that's because the jmx-metrics project overrides the repositories (that's one of gradle quirks that I don't really like that much: `repositories {}` completely replaces everything, there doesn't seem to be a simple way of just appending one particular repo for a single project).
Let's just copy the snapshot repo definition over there for now.

File: static-instrumenter/agent-instrumenter/build.gradle.kts
Comment (Enkelian) [reply static-instrumenter/agent-instrumenter/build.gradle.kts:30]: adding to `DEPENDENCIES` doesn't work so I left it as it is - I'll try placing it there again after the new release

**Review by mateuszrzeszutek — APPROVED** [review]
Thanks! 🚀

**Comment by trask** [comment]
@kubawach it would be great to get your approval as a component owner on this

**Review by kubawach — APPROVED** [review]
Great job!

**Comment by trask** [comment]
thx @Enkelian!

**Pull request merged**
